feat: CLIs for delegation, operator, dogfood

MaxMustermann2 opened the pull request:
- The command line interface to register as an operator has been simplified to rely on labeled arguments instead of positioned arguments.
- It is now possible to supply commission rates while registering an operator and consequently, re-registering has been disabled. An edit function will be implemented in the future to allow edits of the commission rate subject to at most one edit in 24 hours.
- The stateless checks for the operator registration have been moved to `ValidateBasic` as advised by the documentation.
- The CLI interface to opt into or out of an AVS has now been exposed.
- `x/assets/keeper/app_chain.go` has been removed and replaced (for now) with simple checks that `chainID == ctx.ChainID()`.
- The logic between `consensus_keys.go` and `opt.go` within the operator keeper is now shared and dependent. Opting into any Cosmos-based chain is therefore a two step process - first, opt into the AVS and then supply a consensus public key separately. Opting out is the reverse of this.
- To that end, command line transaction methods for setting (or replacing) a consensus public key and removing it (once set) have been introduced. Adjustments for these have been made in the operator hooks implementation in `x/dogfood`.
- Lastly, multiple query commands have been introduced: all operator consensus keys, all operator consensus addresses, single operator consensus address, single operator key, operator opt out finish, undelegation hold count, undelegation maturity and validator.

BREAKING: the format of storing the previous consensus keys has changed.
Previously, they were indexed by operator's AccAddress and chainID. Now
they are indexed by the reverse of it, that is, chainID + AccAddress.
This is done to ensure easy clearing of the data for each specific AVS.

BREAKING: the `prefixSlashAssetsState` and another prefix were sharing
the same value of 5 `BytePrefixForOperatorAndChainIDToConsKey`. This has
been fixed.

MaxMustermann2 committed:
refactor(operator): use flags to RegisterOperator

Instead of relying on the positioning of arguments, the flags mechanism
(as seen in the Cosmos SDK's x/staking) offers a cleaner approach.

MaxMustermann2 committed:
feat(operator): add commission rate to CLI

MaxMustermann2 committed:
refactor(operator): validate registration input

MaxMustermann2 committed:
feat(operator): add `OptIn` to AVS function

MaxMustermann2 committed:
refactor(assets): remove app chains

These should be in the AVS module anyway.

MaxMustermann2 committed:
fix(operator): share `opt.go`, `consensus_keys.go`

An operator must first opt into the chain-as-an-AVS before assigning a
consensus key for said chain. Similarly, to opt out, the operator must
first remove their key and then opt out.

MaxMustermann2 committed:
fix(operator): correct interface

MaxMustermann2 committed:
fix(dogfood): remove unused code

MaxMustermann2 committed:
feat(operator): add AVS opt out CLI

MaxMustermann2 committed:
fix(operator): implement msg server

MaxMustermann2 committed:
feat(operator): implement cons key interfaces

MaxMustermann2 committed:
fix(operator): skip opting out operators from list

MaxMustermann2 committed:
fix typos

MaxMustermann2 committed:
fix(dogfood): update operator hook assumptions

MaxMustermann2 committed:
feat(operator): add QueryAllOperatorKeysByChainID

MaxMustermann2 committed:
feat(operator): implement QueryOperatorConsAddress

MaxMustermann2 committed:
feat(operator): implement QueryAllOperatorConsAddr

MaxMustermann2 committed:
feat(dogfood): implement QueryOptOutsToFinish

MaxMustermann2 committed:
feat(dogfood): QueryOperatorOptOutFinishEpoch

MaxMustermann2 committed:
feat(delegation): query undelegation hold count

MaxMustermann2 committed:
feat(dogfood): implement QueryUndelegationsToMature

MaxMustermann2 committed:
fix(delegation): use string record key type

MaxMustermann2 committed:
feat(dogfood): query undelegation maturity epoch

MaxMustermann2 committed:
feat(dogfood): query validator

MaxMustermann2 committed:
fix(operator): allow build

MaxMustermann2 committed:
fix(dogfood): actually schedule cons addr prune

MaxMustermann2 committed:
chore: lint and tests

bwhour commented on proto/exocore/operator/v1/query.proto:
```diff
@@ -30,18 +31,115 @@ message QueryOperatorConsKeyResponse {
+// OperatorConsAddrPair holds the operator address and the corresponding consensus address.
+message OperatorConsAddrPair {
+  // operator_addr is the operator address.
+  string operator_addr = 1;
```
can settle into one  single style  "operator_addr  cons_addr" or "operator_address  cons_address"

bwhour commented on proto/exocore/operator/v1/query.proto:
```diff
@@ -30,18 +31,115 @@ message QueryOperatorConsKeyResponse {
+// QueryOperatorConsAddressResponse is the response for QueryOperatorConsAddressRequest.
+message QueryOperatorConsAddressResponse {
+  // address is the consensus address of the operator
+  string address = 1;
```
more meaningful naming: cons_addr

bwhour commented on proto/exocore/operator/v1/query.proto:
```diff
@@ -30,18 +31,115 @@ message QueryOperatorConsKeyResponse {
+// for a specific chain ID.
+message QueryOperatorConsAddressRequest {
+  // addr is the ACC address of operator
+  string addr = 1;
```
more meaningful naming: operator_acc_addr ?

bwhour commented on proto/exocore/operator/v1/query.proto:
```diff
@@ -20,8 +21,8 @@ message GetOperatorInfoReq {
-  // chain_id is the id of the chain served by the operator
-  string chain_id = 2;
+  // chain is the id of the chain served by the operator
+  string chain = 2;
```
why  not use chain_id ?

bwhour commented on x/dogfood/client/cli/query.go:
```diff
@@ -2,6 +2,8 @@ package cli
 	"fmt"
+	"strconv"
+
 	// "strings"
```
remove this comment

MaxMustermann2 committed:
fix: use consistent naming of query members

MaxMustermann2 committed:
remove unused comments

MaxMustermann2 committed:
doc: update comment for why chain not chain_id

MaxMustermann2 committed:
Merge branch 'develop' into fix/operator-cli

cloud8little commented:
![47f30f638acc2d06233151f801e0be1](https://github.com/ExocoreNetwork/exocore/assets/34291844/0cec3a8e-91e9-4ac9-af7a-82ed6712c3c5)

cloud8little commented:
**Command**: 

`exocored tx operator register-operator --meta-info operator4 --commission-rate 0 --commission-max-rate 1 --commission-max-change-rate 1 --from testaccount1 --node http://localhost:20000 --fees 50000000000aexo`

**Error Log:** 

```
gas estimate: 152564
auth_info:
  fee:
    amount:
    - amount: "5225469564"
      denom: aexo
    gas_limit: "152564"
    granter: ""
    payer: ""
  signer_infos: []
  tip: null
body:
  extension_options: []
  memo: ""
  messages:
  - '@type': /exocore.operator.v1.RegisterOperatorReq
    from_address: exo1nkh04qzxed98megzz53xx9a4uju599zg3rw0mj
    info:
      approve_addr: exo1nkh04qzxed98megzz53xx9a4uju599zg3rw0mj
      client_chain_earnings_addr:
        earning_info_list: []
      commission:
        commission_rates:
          max_change_rate: "1.000000000000000000"
          max_rate: "1.000000000000000000"
          rate: "0.000000000000000000"
        update_time: "1970-01-01T00:00:00Z"
      earnings_addr: exo1nkh04qzxed98megzz53xx9a4uju599zg3rw0mj
      operator_meta_info: jessica1
  non_critical_extension_options: []
  timeout_height: "0"
signatures: []
confirm transaction before signing and broadcasting [y/N]: y
code: 111222
codespace: undefined
data: ""
events: []
gas_used: "0"
gas_wanted: "0"
height: "0"
info: ""
logs: []
raw_log: "recovered: expected *legacytx.LegacyMsg when using amino JSON\nstack:\ngoroutine
  160346 [running]:\nruntime/debug.Stack()\n\truntime/debug/stack.go:24 +0x5e\ngithub.com/cosmos/cosmos-sdk/baseapp.NewBaseApp.newDefaultRecoveryMiddleware.func3({0x2cbbd00?,
  0xc001a7be40})\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/recovery.go:71 +0x26\ngithub.com/cosmos/cosmos-sdk/baseapp.NewBaseApp.newDefaultRecoveryMiddleware.newRecoveryMiddleware.func5({0x2cbbd00?,
  0xc001a7be40?})\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/recovery.go:39 +0x2d\ngithub.com/cosmos/cosmos-sdk/baseapp.processRecovery({0x2cbbd00,
  0xc001a7be40}, 0x13ff00f0?)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/recovery.go:28
  +0x2f\ngithub.com/cosmos/cosmos-sdk/baseapp.processRecovery({0x2cbbd00, 0xc001a7be40},
  0x0?)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/recovery.go:33 +0x53\ngithub.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runTx.func1()\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/baseapp.go:641
  +0xfe\npanic({0x2cbbd00?, 0xc001a7be40?})\n\truntime/panic.go:920 +0x270\ngithub.com/cosmos/cosmos-sdk/x/auth/ante.SetUpContextDecorator.AnteHandle.func1()\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/auth/ante/setup.go:57
  +0x1e5\npanic({0x2cbbd00?, 0xc001a7be40?})\n\truntime/panic.go:914 +0x21f\ngithub.com/cosmos/cosmos-sdk/x/auth/migrations/legacytx.StdSignBytes({0xc009d2cdf8,
  0x14}, 0xa9, 0x2, 0x0, {{0xc005dad488, 0x1, 0x1}, 0x30d40, {0x0, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/auth/migrations/legacytx/stdsign.go:60
  +0x438\ngithub.com/evmos/evmos/v14/ethereum/eip712.decodeProtobufSignDoc({0xc002aba820,
  0x18b, 0x18b})\n\tgithub.com/evmos/evmos/v14@v14.0.0-rc4/ethereum/eip712/encoding.go:183
  +0x729\ngithub.com/evmos/evmos/v14/ethereum/eip712.GetEIP712TypedDataForMsg({0xc002aba820,
  0x18b, 0x18b})\n\tgithub.com/evmos/evmos/v14@v14.0.0-rc4/ethereum/eip712/encoding.go:60
  +0x251\ngithub.com/evmos/evmos/v14/ethereum/eip712.GetEIP712BytesForMsg({0xc002aba820?,
  0x40?, 0x21?})\n\tgithub.com/evmos/evmos/v14@v14.0.0-rc4/ethereum/eip712/encoding.go:38
  +0x58\ngithub.com/evmos/evmos/v14/crypto/ethsecp256k1.PubKey.verifySignatureAsEIP712({{0xc0066091a0?,
  0xf103fd?, 0xc003d73dd0?}}, {0xc002aba820, 0x18b, 0x18b}, {0xc0030e6b40, 0x41, 0x41})\n\tgithub.com/evmos/evmos/v14@v14.0.0-rc4/crypto/ethsecp256k1/ethsecp256k1.go:222
  +0x65\ngithub.com/evmos/evmos/v14/crypto/ethsecp256k1.PubKey.VerifySignature({{0xc0066091a0?,
  0x4290ed?, 0x40d2b68?}}, {0xc002aba820, 0x18b, 0x18b}, {0xc0030e6b40, 0x41, 0x41})\n\tgithub.com/evmos/evmos/v14@v14.0.0-rc4/crypto/ethsecp256k1/ethsecp256k1.go:215
  +0xaa\ngithub.com/cosmos/cosmos-sdk/x/auth/signing.VerifySignature({0x4108838?,
  0xc005dad440}, {{0xc003d73dd0, 0x2a}, {0xc0075a75d8, 0x14}, 0xa9, 0x2, {0x4108838,
  0xc005dad440}}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/auth/signing/verify.go:21
  +0x10d\ngithub.com/cosmos/cosmos-sdk/x/auth/ante.SigVerificationDecorator.AnteHandle({{_,
  _}, {_, _}}, {{0x40f9158, 0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, ...},
  ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/auth/ante/sigverify.go:292
  +0x63a\ngithub.com/cosmos/cosmos-sdk/types.ChainAnteDecorators.func1({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/types/handler.go:49
  +0x136\ngithub.com/cosmos/cosmos-sdk/x/auth/ante.SigGasConsumeDecorator.AnteHandle({{_,
  _}, _}, {{0x40f9158, 0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8,
  ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/auth/ante/sigverify.go:194
  +0x577\ngithub.com/cosmos/cosmos-sdk/types.ChainAnteDecorators.func1({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/types/handler.go:49
  +0x136\ngithub.com/cosmos/cosmos-sdk/x/auth/ante.ValidateSigCountDecorator.AnteHandle({{_,
  _}}, {{0x40f9158, 0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8,
  0x14}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/auth/ante/sigverify.go:384
  +0x3ce\ngithub.com/cosmos/cosmos-sdk/types.ChainAnteDecorators.func1({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/types/handler.go:49
  +0x136\ngithub.com/cosmos/cosmos-sdk/x/auth/ante.SetPubKeyDecorator.AnteHandle({{_,
  _}}, {{0x40f9158, 0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8,
  0x14}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/auth/ante/sigverify.go:126
  +0xad7\ngithub.com/cosmos/cosmos-sdk/types.ChainAnteDecorators.func1({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/types/handler.go:49
  +0x136\ngithub.com/ExocoreNetwork/exocore/app/ante/cosmos.DeductFeeDecorator.AnteHandle({{0x7f32a80bd438,
  0xc001217900}, {0x7f32a0469a18, 0xc0015e8f20}, {0x40c22a0, 0xc000e2b640}, {0x40c2280,
  0xc001985fb0}, {0x40d1a60, 0xc001945e10}, ...}, ...)\n\tgithub.com/ExocoreNetwork/exocore/app/ante/cosmos/fees.go:88
  +0x397\ngithub.com/cosmos/cosmos-sdk/types.ChainAnteDecorators.func1({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/types/handler.go:49
  +0x136\ngithub.com/cosmos/cosmos-sdk/x/auth/ante.ConsumeTxSizeGasDecorator.AnteHandle({{_,
  _}}, {{0x40f9158, 0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8,
  0x14}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/auth/ante/basic.go:143
  +0x5af\ngithub.com/cosmos/cosmos-sdk/types.ChainAnteDecorators.func1({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/types/handler.go:49
  +0x136\ngithub.com/ExocoreNetwork/exocore/app/ante/cosmos.MinGasPriceDecorator.AnteHandle({{_,
  _}, {_, _}}, {{0x40f9158, 0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, ...},
  ...}, ...)\n\tgithub.com/ExocoreNetwork/exocore/app/ante/cosmos/min_price.go:38
  +0x3d2\ngithub.com/cosmos/cosmos-sdk/types.ChainAnteDecorators.func1({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/types/handler.go:49
  +0x136\ngithub.com/cosmos/cosmos-sdk/x/auth/ante.ValidateMemoDecorator.AnteHandle({{_,
  _}}, {{0x40f9158, 0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8,
  0x14}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/auth/ante/basic.go:67
  +0x1b9\ngithub.com/cosmos/cosmos-sdk/types.ChainAnteDecorators.func1({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/types/handler.go:49
  +0x136\ngithub.com/cosmos/cosmos-sdk/x/auth/ante.TxTimeoutHeightDecorator.AnteHandle({},
  {{0x40f9158, 0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8,
  0x14}, 0x35b45f, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/auth/ante/basic.go:206
  +0x26e\ngithub.com/cosmos/cosmos-sdk/types.ChainAnteDecorators.func1({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/types/handler.go:49
  +0x136\ngithub.com/cosmos/cosmos-sdk/x/auth/ante.ValidateBasicDecorator.AnteHandle({},
  {{0x40f9158, 0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8,
  0x14}, 0x35b45f, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/auth/ante/basic.go:34
  +0x14e\ngithub.com/cosmos/cosmos-sdk/types.ChainAnteDecorators.func1({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/types/handler.go:49
  +0x136\ngithub.com/cosmos/cosmos-sdk/x/auth/ante.RejectExtensionOptionsDecorator.AnteHandle({_},
  {{0x40f9158, 0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8,
  0x14}, 0x35b45f, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/auth/ante/ext.go:52
  +0x10e\ngithub.com/cosmos/cosmos-sdk/types.ChainAnteDecorators.func1({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/types/handler.go:49
  +0x136\ngithub.com/cosmos/cosmos-sdk/x/auth/ante.SetUpContextDecorator.AnteHandle({},
  {{0x40f9158, 0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8,
  0x14}, 0x35b45f, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/x/auth/ante/setup.go:62
  +0x56e\ngithub.com/cosmos/cosmos-sdk/types.ChainAnteDecorators.func1({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/types/handler.go:49
  +0x136\ngithub.com/ExocoreNetwork/exocore/app/ante/cosmos.AuthzLimiterDecorator.AnteHandle({{_,
  _, _}}, {{0x40f9158, 0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8,
  ...}, ...}, ...}, ...)\n\tgithub.com/ExocoreNetwork/exocore/app/ante/cosmos/authz.go:33
  +0x18e\ngithub.com/cosmos/cosmos-sdk/types.ChainAnteDecorators.func1({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/types/handler.go:49
  +0x136\ngithub.com/ExocoreNetwork/exocore/app/ante/cosmos.RejectMessagesDecorator.AnteHandle({},
  {{0x40f9158, 0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8,
  0x14}, 0x35b45f, ...}, ...}, ...)\n\tgithub.com/ExocoreNetwork/exocore/app/ante/cosmos/reject_msgs.go:25
  +0x14e\ngithub.com/cosmos/cosmos-sdk/types.ChainAnteDecorators.func1({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/types/handler.go:49
  +0x136\ngithub.com/ExocoreNetwork/exocore/app.(*ExocoreApp).setAnteHandler.NewAnteHandler.func2({{0x40f9158,
  0x5f8e340}, {0x4114290, 0xc008926000}, {{0xb, 0x0}, {0xc0075a75d8, 0x14}, 0x35b45f,
  {0x2286648f, ...}, ...}, ...}, ...)\n\tgithub.com/ExocoreNetwork/exocore/app/ante/ante.go:53
  +0x218\ngithub.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runTx(0xc000320d20, 0x0,
  {0xc002ad0a80, 0x1b5, 0x1b6})\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/baseapp.go:696
  +0x4f9\ngithub.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).CheckTx(0xc000320d20, {{0xc002ad0a80?,
  0x868940c709eefc85?, 0x4ed8aa4a391c0cb3?}, 0x5b9cca4f?})\n\tgithub.com/cosmos/cosmos-sdk@v0.47.5/baseapp/abci.go:370
  +0x86\ngithub.com/cometbft/cometbft/abci/client.(*localClient).CheckTxAsync(0xc001a00600,
  {{0xc002ad0a80?, 0x0?, 0x0?}, 0x307cbd8?})\n\tgithub.com/cometbft/cometbft@v0.37.2/abci/client/local_client.go:93
  +0x102\ngithub.com/cometbft/cometbft/proxy.(*appConnMempool).CheckTxAsync(0xc000df65a0,
  {{0xc002ad0a80?, 0x548c3d?, 0xc00307cc68?}, 0x432c05?})\n\tgithub.com/cometbft/cometbft@v0.37.2/proxy/app_conn.go:154
  +0x171\ngithub.com/cometbft/cometbft/mempool/v0.(*CListMempool).CheckTx(0xc000e4fd40,
  {0xc002ad0a80, 0x1b5, 0x1b6}, 0xc005dad008, {0x20?, {0x0?, 0x50003?}})\n\tgithub.com/cometbft/cometbft@v0.37.2/mempool/v0/clist_mempool.go:254
  +0x1c3\ngithub.com/cometbft/cometbft/rpc/core.BroadcastTxSync(0xc0068d6ae0, {0xc002ad0a80,
  0x1b5, 0x1b6})\n\tgithub.com/cometbft/cometbft@v0.37.2/rpc/core/mempool.go:35 +0xe4\nreflect.Value.call({0x2cfa080?,
  0x3ae68d8?, 0x3195360?}, {0x31a7644, 0x4}, {0xc005d58150, 0x2, 0x439572?})\n\treflect/value.go:596
  +0xce7\nreflect.Value.Call({0x2cfa080?, 0x3ae68d8?, 0x251?}, {0xc005d58150?, 0x3195360?,
  0xc0032ebc00?})\n\treflect/value.go:380 +0xb9\ngithub.com/cometbft/cometbft/rpc/jsonrpc/server.RegisterRPCFuncs.makeJSONRPCHandler.func3({0x40e95b0,
  0xc005dacf60}, 0xc004f87300)\n\tgithub.com/cometbft/cometbft@v0.37.2/rpc/jsonrpc/server/http_json_handler.go:108
  +0xcf0\ngithub.com/cometbft/cometbft/rpc/jsonrpc/server.RegisterRPCFuncs.handleInvalidJSONRPCPaths.func4({0x40e95b0?,
  0xc005dacf60?}, 0x7f32a8041380?)\n\tgithub.com/cometbft/cometbft@v0.37.2/rpc/jsonrpc/server/http_json_handler.go:140
  +0x42\nnet/http.HandlerFunc.ServeHTTP(0xc000600000?, {0x40e95b0?, 0xc005dacf60?},
  0xc00307d968?)\n\tnet/http/server.go:2136 +0x29\nnet/http.(*ServeMux).ServeHTTP(0xc000600000?,
  {0x40e95b0, 0xc005dacf60}, 0xc004f87300)\n\tnet/http/server.go:2514 +0x142\ngithub.com/cometbft/cometbft/rpc/jsonrpc/server.maxBytesHandler.ServeHTTP({{0x40c3660?,
  0xc0005ba180?}, 0xd?}, {0x40e95b0?, 0xc005dacf60}, 0xc004f87300)\n\tgithub.com/cometbft/cometbft@v0.37.2/rpc/jsonrpc/server/http_server.go:256
  +0x11c\ngithub.com/cometbft/cometbft/rpc/jsonrpc/server.Serve.RecoverAndLogHandler.func1({0x40df760?,
  0xc000916540}, 0xc004f87300)\n\tgithub.com/cometbft/cometbft@v0.37.2/rpc/jsonrpc/server/http_server.go:229
  +0x26f\nnet/http.HandlerFunc.ServeHTTP(0x432f85?, {0x40df760?, 0xc000916540?}, 0xc000916501?)\n\tnet/http/server.go:2136
  +0x29\nnet/http.serverHandler.ServeHTTP({0x40d2398?}, {0x40df760?, 0xc000916540?},
  0x6?)\n\tnet/http/server.go:2938 +0x8e\nnet/http.(*conn).serve(0xc006cb3710, {0x40f9660,
  0xc004fc3b60})\n\tnet/http/server.go:2009 +0x5f4\ncreated by net/http.(*Server).Serve
  in goroutine 97\n\tnet/http/server.go:3086 +0x5cb\n: panic"
```

MaxMustermann2 committed:
fix(operator): add Router for tx

MaxMustermann2 commented:
The latest commit fixes it.

Please ensure to supply lots of gas with `--gas auto --gas-adjustment 2.0 --gas-prices 34251aexo`

MaxMustermann2 committed:
chore(lint)

MaxMustermann2 committed:
Merge develop into fix/operator-cli

TimmyExogenous commented on x/operator/keeper/operator.go:
```diff
@@ -3,28 +3,62 @@ package keeper
+	ctx sdk.Context, addr string, info *operatortypes.OperatorInfo,
+) (err error) {
+	// the operator's `addr` must match the earnings address.
+	if addr != info.EarningsAddr {
```
Should we enforce the operator address as the earning address? Most blockchains do not use the same address for rewards and operations. They utilize separate addresses to mitigate the risk of private key loss. This approach may enhance security for operations and maintenance, as only the worker address needs to be stored on the server.

MaxMustermann2 replied:
Which other address can we use?

MaxMustermann2 committed:
fix(test): fix operator tests, disable 1

TimmyExogenous commented on x/operator/keeper/opt.go:
```diff
@@ -17,9 +18,11 @@ type AssetPriceAndDecimal struct {
-		return types.ErrInvalidAvsAddr
+	// avsAddr should be an evm contract address or a chain id.
+	if !common.IsHexAddress(avsAddr) {
+		if avsAddr != ctx.ChainID() { // TODO: other chain ids besides this chain's.
```
Should we allocate a default EVM contract address for the dogfood to make the addresses of all AVSs are the same type?  It can be implemented in the AVS module.

MaxMustermann2 replied:
Yes, that would make sense.

MaxMustermann2 committed:
fix(proto): prevent complains about duplicate reg

MaxMustermann2 commented:
> ![47f30f638acc2d06233151f801e0be1](https://private-user-images.githubusercontent.com/34291844/326651997-0cec3a8e-91e9-4ac9-af7a-82ed6712c3c5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ0NjY3NjksIm5iZiI6MTcxNDQ2NjQ2OSwicGF0aCI6Ii8zNDI5MTg0NC8zMjY2NTE5OTctMGNlYzNhOGUtOTFlOS00YWM5LWFmN2EtODJlZDY3MTJjM2M1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDMwVDA4NDEwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIxZWYyZTFlZTUyNzcyMTlhNDRhODkzNjg0ZThkNDNjNTA1NGVjMzNiOTRjOTJhZWRiMjJhOTM4ZjIwZWI1MDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1nwl-AKNxOgh12UJUwXqFS3iAS3AvmWKqkWeuKMFhyw)

The last commit resolves this error as well.

MaxMustermann2 committed:
fix(operator): query operator CLI with address

TimmyExogenous commented on x/operator/keeper/opt.go:
```diff
@@ -55,10 +58,28 @@ func (k *Keeper) OptIn(ctx sdk.Context, operatorAddress sdk.AccAddress, avsAddr
+		if avsAddr == ctx.ChainID() {
+			found, _ := k.getOperatorConsKeyForChainID(ctx, operatorAddress, avsAddr)
+			if found {
+				// if the key exists, it should be in the process of being removed.
```
For the dogfood, the entry process should be:
1. set consensus key
2. opt-in

The exit process is:
1. opt-out
2. stopping the node and removing the consensus key after the voting power is updated at the end of the epoch

So why is the key needed to be in the process of being removed here?

TimmyExogenous commented on x/operator/keeper/opt.go:
```diff
@@ -55,10 +58,28 @@ func (k *Keeper) OptIn(ctx sdk.Context, operatorAddress sdk.AccAddress, avsAddr
+			if found {
+				// if the key exists, it should be in the process of being removed.
+				// TODO: if slashing is moved to a snapshot approach, opt out should only be
+				// performed if the key doesn't exist.
```
Why is it going to be influenced by the slash mechanism here?

leonz789 commented on x/operator/client/cli/tx.go:
```diff
@@ -1,81 +1,228 @@
+}
+
+func newBuildRegisterOperatorMsg(
+	clientCtx client.Context, txf tx.Factory, fs *flag.FlagSet,
```
`txf` unused expect for return value

cloud8little commented:
Tested 5099fcb8a5b2d9259c0798488e4307be66030c2a,  still gets the error here https://github.com/ExocoreNetwork/exocore/pull/46#issuecomment-2084582779. 

More investigation:
This did not happen on single node running a new chain, happens on the export testnet V1 data and import bootstrap data, and three nodes consensus.

cloud8little commented:
After register operator, it is ok to query operator info with `get-operator-info`, but not able to query consensus key `get-operator-cons-key` and consensus address `get-operator-cons-address`, and there is no info for the new added operator in `get-all-operator-cons-addrs` and `get-all-operators-by-chain-id`

```
[root@testnet-exochain ~]# exocored query operator get-operator-info exo16hgwvex9jm4ztxpf2d7dee56qd50tch2a84c6a --home ~/.tmp-exocored/
approve_addr: exo16hgwvex9jm4ztxpf2d7dee56qd50tch2a84c6a
client_chain_earnings_addr:
  earning_info_list: []
commission:
  commission_rates:
    max_change_rate: "1.000000000000000000"
    max_rate: "1.000000000000000000"
    rate: "0.000000000000000000"
  update_time: "2024-05-06T10:07:45.065045865Z"
earnings_addr: exo16hgwvex9jm4ztxpf2d7dee56qd50tch2a84c6a
operator_meta_info: operator4


[root@testnet-exochain ~]# exocored query operator get-operator-cons-address exo16hgwvex9jm4ztxpf2d7dee56qd50tch2a84c6a exocoretestnet_233-1 --home ~/.tmp-exocored/
Error: rpc error: code = InvalidArgument desc = no key assigned: invalid request

exocored query operator get-operator-cons-key exo16hgwvex9jm4ztxpf2d7dee56qd50tch2a84c6a exocoretestnet_233-1 --home ~/.tmp-exocored
Error: rpc error: code = InvalidArgument desc = no key assigned: invalid request

[root@testnet-exochain ~]# exocored query operator get-all-operator-cons-addrs exocoretestnet_233-1 --home ~/.tmp-exocored/
operator_cons_addrs:
- cons_addr: exovalcons18z3p42xn8pjk338upvzp794h02wh7p4t7jj9jx
  operator_acc_addr: exo18cggcpvwspnd5c6ny8wrqxpffj5zmhklprtnph
pagination:
  next_key: null
  total: "1"
[root@testnet-exochain ~]# exocored query operator get-all-operators-by-chain-id exocoretestnet_233-1 --home ~/.tmp-exocored/
operator_cons_keys:
- operator_acc_addr: exo18cggcpvwspnd5c6ny8wrqxpffj5zmhklprtnph
  public_key:
    ed25519: 8PaRnlIsW5fbLIJVv/dD+d/d162fw3ywwWcLSA0PmRQ=
pagination:
  next_key: null
  total: "1"

```

MaxMustermann2 committed:
feat(operator): add set-cons-key and removal cli

MaxMustermann2 committed:
fix(dogfood,operator): clear prev keys correctly

AVSs are free to choose their own epoch schedule. The previous key needs
to be cleared only at epoch end (and not at block end as it was done
previously). Additionally, it is specific to the AVS identifier (the
chainID in the case of the dogfood module) that should be cleared.

BREAKING: the format of storing the previous consensus keys has changed.
Previously, they were indexed by operator's AccAddress and chainID. Now
they are indexed by the reverse of it, that is, chainID + AccAddress.
This is done to ensure easy clearing of the data for each specific AVS.

BREAKING: the `prefixSlashAssetsState` and another prefix were sharing
the same value of 5 `BytePrefixForOperatorAndChainIDToConsKey`. This has
been fixed.

MaxMustermann2 committed:
fix(operator): remove pointer receiver

MaxMustermann2 commented:
BREAKING: the format of storing the previous consensus keys has changed.
Previously, they were indexed by operator's AccAddress and chainID. Now
they are indexed by the reverse of it, that is, chainID + AccAddress.
This is done to ensure easy clearing of the data for each specific AVS.

BREAKING: the `prefixSlashAssetsState` and another prefix were sharing
the same value of 5 `BytePrefixForOperatorAndChainIDToConsKey`. This has
been fixed.

MaxMustermann2 committed:
refactor: clarify all vs active operators

cloud8little commented:
Tested 95bde9820c246ab750e3e0071eb8a1453528ceb9, the issue still exists.

cloud8little commented:
Test binary: https://github.com/ExocoreNetwork/exocore/pull/46/commits/95bde9820c246ab750e3e0071eb8a1453528ceb9
1. register operator with from account dev0.
2. `exocored tx operator opt-into-avs exocoretestnet_233-1 --from dev0 --home ~/.tmp-exocored/ --gas-prices 5000000aexo`
3. `exocored tx operator set-cons-key exocoretestnet_233-1 '{"@type":"/cosmos.crypto.ed25519.PubKey","key":"LlzL2PWhXBgDyTuK1qNXhHxy3+dpcq8ZkT9/GSY6318="}' --from dev0 --home ~/.tmp-exocored --gas-prices 5000000aexo`
4. query transaction.
5. error log. 
[set_cons_key_errlog.txt](https://github.com/ExocoreNetwork/exocore/files/15250870/set_cons_key_errlog.txt)

MaxMustermann2 committed:
feat(operator): add get all operators query

MaxMustermann2 committed:
fix(operator): parse cons key correctly

cloud8little commented:
Retesting.

MaxMustermann2 committed:
Merge develop

MaxMustermann2 committed:
chore(lint): placate proto-lint

MaxMustermann2 committed:
chore(lint): placate golang lint

cloud8little commented:
Test Result For https://github.com/ExocoreNetwork/exocore/pull/46/commits/eb2959c17f5ebd7440ce2f0bbbb3e40da114d43d， test scenario:

1. Register Operator for a new operator with a non existing operator name.  [**Passed**]
2. Register Operator for a new operator with an existing operator name.  [**Passed**]
3. Register operator for the same account for several times.   [**Passed**]
4. Set consensus key for the new registered operator without opt into avs.   [**Passed**]
5. Set consensus key for the new registered operator after opting into avs.    [**Passed**] 
6. Set consensus key for the new registered operator after opting into avs the second time.   [**Passed**]
7. Check all opeartors by chain id after register the new operator but without setting a consensus key   [**Passed**]
8. Check the new registered operator consensus key info and operator info without setting a consensus key.   [**Passed**]
9. Check the new registered operator consensus key info and operator info after setting a consensus key.   [**Passed**]
10. Register operator, opt into avs, set consensus key, then initiate key removal and opt-out avs. check the specific operator info and all operator info. [The consensus key still shows ]

TimmyExogenous commented on x/dogfood/client/cli/query.go:
```diff
@@ -60,3 +62,151 @@ func CmdQueryParams() *cobra.Command {
+	return cmd
+}
+
+func CmdQueryUndelegationsToMature() *cobra.Command {
```
Should we still provide these functions in dogfood, seems not necessary here. All undelegations information should be retrieved from the `delegation` module, right?

MaxMustermann2 replied:
This function lists out the undelegations that will mature at epoch N. That information is not available with the delegation module.

TimmyExogenous commented on x/dogfood/client/cli/query.go:
```diff
@@ -60,3 +62,151 @@ func CmdQueryParams() *cobra.Command {
 	return cmd
 }
+
+func CmdQueryOptOutsToFinish() *cobra.Command {
```
As we discussed before, the `opt_out` might not depend on the maturity of the undelegations, right? So should we remove these functions?

MaxMustermann2 replied:
I am still confused. We should speak once. But there is certainly some refactoring here that can be done.

MaxMustermann2 commented:
[errolog_register.txt](https://github.com/ExocoreNetwork/exocore/files/15339937/errolog_register.txt)

I cannot reproduce this error. If it doesn't stop consensus, it looks to be an issue in the sender-side of things, especially based on the logs which do not mention anything related to the modules.

![image](https://github.com/ExocoreNetwork/exocore/assets/82761650/27aa70ee-520e-4595-af69-bbdc078ccc48)

MaxMustermann2 commented:
> Register operator, opt into avs, set consensus key, then initiate key removal and opt-out avs. check the specific operator info and all operator info. [The consensus key still shows ]

That is intentional. It stays till maturity of the unbonding period. However, let me discuss with Timmy again - based on the slashing design this will change.

MaxMustermann2 commented:
> I cannot reproduce this error. If it doesn't stop consensus, it looks to be an issue in the sender-side of things, especially based on the logs which do not mention anything related to the modules.

I am able to reproduce this error by providing the wrong chain id to the call. This means it is a problem of the sender and not the receiver and is not a cause for concern. It does not stop consensus at all.

The same doesn't hold for `tx bank send` which reports a nice clean error. Let me see if we can get that.

MaxMustermann2 committed:
fix(operator): impl for msgs GetSignBytes, Type

Without these functions implemented, sending a transaction with the
wrong chain-id produces a panic while verifying the signature. This
panic happens on the receiving node (and not the sender), although
consensus isn't impacted.

With these functions implemented, the receiving node replies "signature
verification failed, please verify account number (%d) and chain-id
(%s)". If the transaction were only legacy amino, even the "sequence
(%d)" would have shown up in the error message.

cloud8little commented:
Test object: https://github.com/ExocoreNetwork/exocore/pull/46/commits/11a8a960af2947c6e4a49de16cd9e968b917c030

Scenario 1.  deployer1 account is under default folder ~/.exocored and no balances, shows `key not found `error.

```shell
exocored tx operator register-operator --meta-info deployer1 --commission-rate 0 --commission-max-rate 1 --commission-max-change-rate 1 --from deployer1 --node http://localhost:20000 --gas-prices 7aexo
Enter keyring passphrase (attempt 1/3):
Error: rpc error: code = NotFound desc = rpc error: code = NotFound desc = account exo1hn2qgt0yn8g5u4gqrn9myjj4ruae2syk76pvk5 not found: key not found
```

Scenario 2.  testaccount1 has balances, and is under default folder ~/.exocored. register operator successfully.

```shell
exocored tx operator register-operator --meta-info testoperator1 --commission-rate 0 --commission-max-rate 1 --commission-max-change-rate 1 --from testaccount1 --node http://localhost:20000 --gas-prices 7aexo
Enter keyring passphrase (attempt 1/3):
auth_info:
  fee:
    amount:
    - amount: "1400000"
      denom: aexo
    gas_limit: "200000"
    granter: ""
    payer: ""
  signer_infos: []
  tip: null
body:
  extension_options: []
  memo: ""
  messages:
  - '@type': /exocore.operator.v1.RegisterOperatorReq
    from_address: exo1nkh04qzxed98megzz53xx9a4uju599zg3rw0mj
    info:
      approve_addr: exo1nkh04qzxed98megzz53xx9a4uju599zg3rw0mj
      client_chain_earnings_addr:
        earning_info_list: []
      commission:
        commission_rates:
          max_change_rate: "1.000000000000000000"
          max_rate: "1.000000000000000000"
          rate: "0.000000000000000000"
        update_time: "1970-01-01T00:00:00Z"
      earnings_addr: exo1nkh04qzxed98megzz53xx9a4uju599zg3rw0mj
      operator_meta_info: testoperator1
  non_critical_extension_options: []
  timeout_height: "0"
signatures: []
confirm transaction before signing and broadcasting [y/N]: y
code: 0
codespace: ""
data: ""
events: []
gas_used: "0"
gas_wanted: "0"
height: "0"
info: ""
logs: []
raw_log: '[]'
timestamp: ""
tx: null
txhash: A89DC8B45AB57A87B309CE2FC6A3913B9779831C8EF7C5FBA2B8F4A18C51322C
```

```shell
exocored query operator get-operator-info exo1nkh04qzxed98megzz53xx9a4uju599zg3rw0mj --n
ode http://localhost:20000
approve_addr: exo1nkh04qzxed98megzz53xx9a4uju599zg3rw0mj
client_chain_earnings_addr:
  earning_info_list: []
commission:
  commission_rates:
    max_change_rate: "1.000000000000000000"
    max_rate: "1.000000000000000000"
    rate: "0.000000000000000000"
  update_time: "2024-05-23T08:52:11.854962787Z"
earnings_addr: exo1nkh04qzxed98megzz53xx9a4uju599zg3rw0mj
operator_meta_info: testoperator1
```
